src/typesystem.cpp

```diff
@@ -191,3 +98,109 @@ void TypeSystemCore::gather_free_vars(MonoId mono, std::unordered_set<MonoId>& f
 	}
 }
+
+
+
+TypeFunctionId TypeSystemCore::new_builtin_type_function(int arity) {
+	return create_tf(TypeFunctionTag::Builtin, arity, {}, {}, false);
+}
+
+TypeFunctionId TypeSystemCore::new_type_function(
+    TypeFunctionTag type,
+    std::vector<InternedString> fields,
+    std::unordered_map<InternedString, MonoId> structure,
+    bool dummy) {
+	return create_tf(type, 0, std::move(fields), std::move(structure), dummy);
+}
+
+TypeFunctionId TypeSystemCore::new_tf_var() {
+	return create_tf(TypeFunctionTag::Builtin, -1, {}, {}, true);
+}
+
+TypeFunctionId TypeSystemCore::create_tf(
+    TypeFunctionTag tag,
+    int arity,
+    std::vector<InternedString> fields,
+    std::unordered_map<InternedString, MonoId> structure,
+    bool is_dummy) {
+	TypeFunctionId result = m_tf_uf.new_var();
+	m_type_functions.push_back(
+	    {tag, arity, std::move(fields), std::move(structure), is_dummy});
+	return result;
+}
+
+TypeFunctionData& TypeSystemCore::type_function_data_of(MonoId mono){
+	TypeFunctionId tf = m_mono_core.find_function(mono);
+	return get_tf_data(tf);
+}
+
+void TypeSystemCore::unify_tf(TypeFunctionId i, TypeFunctionId j) {
+	i = find_tf(i);
+	j = find_tf(j);
+
+	if (i == j)
+		return;
+
+	if (get_tf_data(j).is_dummy)
+		std::swap(i, j);
+
+	if (get_tf_data(i).is_dummy) {
+		point_tf_at_another(i, j);
+		unify_tf_data(get_tf_data(i), get_tf_data(j));
+	} else {
+		Log::fatal() << "unified different typefuncs";
+	}
+}
+
+void TypeSystemCore::point_tf_at_another(TypeFunctionId a, TypeFunctionId b) {
+	m_tf_uf.join_left_to_right(a, b);
+}
+
+int TypeSystemCore::compute_new_argument_count(
+    TypeFunctionData const& a_data, TypeFunctionData const& b_data) const {
+	if (a_data.argument_count == b_data.argument_count) {
+		return a_data.argument_count;
+	} else if (b_data.is_dummy || b_data.argument_count == -1) {
+		return -1;
+	} else {
+		std::string argc_a = a_data.argument_count == -1
+			? std::string("variadic")
+			: std::to_string(a_data.argument_count);
+		std::string argc_b = b_data.argument_count == -1
+			? std::string("variadic")
+			: std::to_string(b_data.argument_count);
+		Log::fatal()
+			<< "Deduced type functions with incompatible argument "
+			"counts to be equal (with "
+			<< argc_a << " and " << argc_b << " arguments)";
+	}
+}
+
+void TypeSystemCore::unify_tf_data(TypeFunctionData& a_data, TypeFunctionData& b_data) {
+
+	int const new_argument_count = compute_new_argument_count(a_data, b_data);
+
+	b_data.argument_count = new_argument_count;
+
+	for (auto& kv_a : a_data.structure) {
+		auto kv_b = b_data.structure.find(kv_a.first);
+
+		if (kv_b == b_data.structure.end())
+			// if b doesn't have a field of a, act accordingly
+			if (b_data.is_dummy)
+				b_data.structure.insert(kv_a);
+			else
+				Log::fatal() << "Accessing non-existing field '" << kv_a.first << "' of a record";
+		else
+			// else the fields must have equivalent types
+			m_mono_core.unify(kv_a.second, kv_b->second);
```

Comment:
Not very important but I believe flattening this branch logic improves readability. You can choose to keep it like this.

Reply:
aight I'll try some stuff out